### PR TITLE
Use new winstone method to run on a random port

### PR DESCRIPTION
**Pull request by Vlatombe**
* Downstream of https://github.com/jenkinsci/winstone/pull/174

Detects Jenkins version:
* For 2.339 and above, it starts winstone on port 0 (random port allocated by jetty) and captures the port using the new winstone option. The port is stable across restarts through the JenkinsController.
* Before 2.339, we keep the previous logic which has risks of port collisions.

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java [hunk 1]
```diff
@@ -43,27 +49,55 @@ public void addJavaOpt(String javaOpt) {
         }
     }
 
-    protected final int httpPort;
+    protected int httpPort;
 
     @Inject
     public WinstoneController(Injector i) {
-        this(i, IOUtil.randomTcpPort());
+        this(i, IOUtil.randomTcpPort()); // TODO Change to 0 when Jenkins has https://github.com/jenkinsci/jenkins/pull/5928
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:56]: https://github.com/jenkinsci/jenkins/pull/5928

File: src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java [hunk 2]
```diff
@@ -43,27 +49,55 @@ public void addJavaOpt(String javaOpt) {
...
+        } catch (NumberFormatException e) {
+            // Should not happen assuming this is called after a successful start.
+            throw new IOException("Unable to parse port from " + s + ". Jenkins did not start.");
+        }
+    }
+
+
     @Override
     public ProcessInputStream startProcess() throws IOException{
+        portFile = File.createTempFile("jenkins-port", ".txt");
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:92]: when is this file cleaned up?

**Review by jtnord — COMMENTED** [review]
LGTM - but we should tidy up the test files on shutdown  / stop

File: src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java [hunk 3]
```diff
@@ -43,27 +49,55 @@ public void addJavaOpt(String javaOpt) {
...
+
+    @Override
+    protected void onReady() throws IOException {
+        if (this.httpPort == 0) {
+            this.httpPort = readPort(portFile);
+        }
+    }
+
+    private static int readPort(File portFile) throws IOException {
+        String s = FileUtils.readFileToString(portFile, StandardCharsets.UTF_8);
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:80]: incorrect charset is being used here?
https://github.com/jenkinsci/winstone/pull/174#discussion_r752472871

Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:80]: `[0-9]+` so who cares?

Comment (jtnord) [reply src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:80]: some people extend the ATH to test their workflows and and some people use that funky IBM thingy....
I have no idea if those users overlap or not - but this is a trivial thing to fix?

Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:80]: z/OS only for agents I think?

File: src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java (hunk 2 above)
Comment (Vlatombe) [reply src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:92]: fixed on next line

**Comment by Vlatombe** [comment]
Not working for test cases restarting Jenkins by themselves. Since winstone is started with `httpPort=0`, the value changes after restart, and this is not being captured by the `JenkinsController` instance.

**Comment by timja** [comment]
> Not working for test cases restarting Jenkins by themselves. Since winstone is started with `httpPort=0`, the value changes after restart, and this is not being captured by the `JenkinsController` instance.

Those are the bane of my existence when fixing tests.

restart doesn’t work on Mac M1

so I have to manually start it up with the correct flags when working on those tests :(

**Comment by Vlatombe** [comment]
@timja I've applied another restart method which shouldn't be dependant anymore on the lifecycle.

**Comment by timja** [comment]
> @timja I've applied another restart method which shouldn't be dependant anymore on the lifecycle.

Thanks!

**Comment by timja** [comment]
> Looks like it works overall. There seems to be some test failures due to recent changes to the credentials plugin.

need fixing before integrating any changes on top of it. Otherwise we end up possibly making the situation worse

File: src/main/java/org/jenkinsci/test/acceptance/controller/LocalController.java [hunk 4]
```diff
@@ -225,14 +225,22 @@ public void startNow() throws IOException {
         logWatcher = new JenkinsLogWatcher(getLogId(),process,logFile, getLogPrinter());
         logWatcher.start();
         try {
-            LOGGER.info("Waiting for Jenkins to become running in "+ this);
+            LOGGER.info("Waiting for Jenkins to become running "+ this);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/controller/LocalController.java:228]: ?

File: src/main/java/org/jenkinsci/test/acceptance/controller/LocalController.java [hunk 5]
```diff
@@ -225,14 +225,22 @@ public void startNow() throws IOException {
...
             diagnoseFailedLoad(e);
         }
     }
 
+    /**
+     * Called when the Jenkins instance is ready to be used.
+     */
+    protected void onReady() throws IOException {
+
+    }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/controller/LocalController.java:242]:
```suggestion
    protected void onReady() throws IOException {}
```

File: src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java [hunk 6]
```diff
@@ -43,29 +51,63 @@ public void addJavaOpt(String javaOpt) {
...
+        if (this.httpPort == 0 && portFile != null) {
+            String s = FileUtils.readFileToString(portFile, StandardCharsets.UTF_8);
+            try {
+                return Integer.parseInt(s);
+            } catch (NumberFormatException e) {
+                // Should not happen assuming this is called after a successful start.
+                throw new IOException("Unable to parse port from " + s + ". Jenkins did not start.");
+            }
+        } else {
+            return this.httpPort;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java:92]: A bit confusing. Perhaps this method should return `void` and edit `[this.]httpPort` only when necessary.

File: src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java [hunk 7]
```diff
@@ -164,21 +166,17 @@ public PluginManager getPluginManager() {
...
 
-    /** 
-     * Some tests require they restart Jenkins - but depending on how the SUT is launched this is not always possible
-     * so tests that require this should wrap this call in an {@link org.junit.Assume#assumeTrue(String, boolean)}
-     * @return true if the Jenkins under test can restart itself.
-     */
     public boolean canRestart() {
-        visit("restart");
-        return getElement(by.button("Yes")) != null;
+        return true;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java:170]: Unclear from 7c40dfb37529ee1d9a65ff48af9c46712b238706 why this is being done here.

**Review by jglick — APPROVED** [review]
OK I think. Not very familiar with ATH at this point.

File: src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java (hunk 7 above)
Comment (Vlatombe) [reply src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java:170]: https://github.com/jenkinsci/acceptance-test-harness/pull/711#issuecomment-1080647705

Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java:170]: So IIUC this now uses https://github.com/jenkinsci/acceptance-test-harness/blob/a9027d40be247fbc51f4ea4af7e4bd0bd7f7e6e8/src/main/java/org/jenkinsci/test/acceptance/controller/JenkinsController.java#L148-L156 which ignores the `hudson.Lifecycle` (and thus the crazy JNI hacks) and just shuts down and restarts the controller using the external system? Like https://github.com/jenkinsci/jenkins/pull/5899#issuecomment-999660881

Can we deprecate `canRestart` and delete existing calls?

**Review by jglick — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Review by jtnord — APPROVED** [review]

**Comment by jtnord** [comment]
@Vlatombe is this ready to go?

**Comment by timja** [comment]
@Vlatombe  ?

**Pull request merged**